### PR TITLE
MMCA-4851 | Implement search by MRN/UCR and View transactions on the Cash Account dashboard

### PR DESCRIPTION
**Pull request by Neozxc**
Tasks:
- Read the value entered in the search box
- Determine the search type based on regex for MRN/UCR and Payment type
- On clicking search button, trigger ACC44 request with a relevant request for MRN and UCR 
- Display the search results page on a successful response

### CODE REVIEW

File: app/controllers/DeclarationDetailController.scala [hunk 1]
```diff
@@ -41,41 +43,109 @@ class DeclarationDetailController @Inject()(authenticate: IdentifierAction,
...
-
+          retrieveCashAccountTransactionAndDisplay(account, from, to, ref, page, cameViaSearch, searchInput)
         case None => Future.successful(NotFound(errorHandler.notFoundTemplate))
       }
+    } yield result
+  }
+
+  def handleSearchRequest(page: Option[Int], searchInput: String)
+                         (implicit request: IdentifierRequest[AnyContent]): Future[Result] = {
+    if (isValidMRNUCR(searchInput) || isValidPayment(searchInput)) {
```
Comment (AdamCoops) [inline app/controllers/DeclarationDetailController.scala:72]: do we want a new line between the = { and the first if statement for readability?

File: app/controllers/DeclarationDetailController.scala [hunk 2]
```diff
@@ -41,41 +43,109 @@ class DeclarationDetailController @Inject()(authenticate: IdentifierAction,
...
+    } else {
+      val paymentAmount = parsePaymentAmount(searchInput)
+      (declaration: Declaration) => declaration.amount == paymentAmount
+    }
+  }
+
+  def isValidMRNUCR(value: String): Boolean =
+    mrnUCRRegex.findFirstIn(value).isDefined || superMRNUCRRegex.findFirstIn(value).isDefined
+
+  def isValidPayment(value: String): Boolean =
```
Comment (AdamCoops) [inline app/controllers/DeclarationDetailController.scala:94]: I don't think this needs to be two separate lines, although I appreciate the consistency as it matches the other defs. Not sure if we just want to move it two one or not?

File: app/utils/Utils.scala [hunk 3]
```diff
@@ -30,6 +30,7 @@ object Utils {
   val hyphen = "-"
   val singleSpace = " "
   val negativeSign = "-"
+  val poundSymbol = "£"
```
Comment (AdamCoops) [inline app/utils/Utils.scala:33]: Can we add a test to UtilsSpec to cover these changes?

File: app/utils/Utils.scala [hunk 4]
```diff
@@ -38,6 +39,7 @@ object Utils {
   val emptyHmrcNewTabLink: HmrcNewTabLink = new HmrcNewTabLink()
   val emptyGovUkTableComponent = new GovukTable()
   val emptyH2InnerComponent = new h2Inner()
+  val emptyH1InnerComponent = new h1Inner()
```
Comment (AdamCoops) [inline app/utils/Utils.scala:42]: and these

File: test/controllers/CashAccountV2ControllerSpec.scala [hunk 5]
```diff
@@ -29,19 +41,16 @@ import play.api.test.Helpers.*
 import services.AuditingService
 import uk.gov.hmrc.http.UpstreamErrorResponse
 import utils.SpecBase
-import views.html.{
-  cash_account_no_transactions_v2, cash_account_no_transactions_with_balance,
-  cash_account_transactions_not_available
-}
+import views.html.{cash_account_no_transactions_v2, cash_account_no_transactions_with_balance, cash_account_transactions_not_available}
```
Comment (AdamCoops) [inline test/controllers/CashAccountV2ControllerSpec.scala:44]: line length is too long

File: test/controllers/DeclarationDetailControllerSpec.scala
Comment (AdamCoops) [inline test/controllers/DeclarationDetailControllerSpec.scala:1]: New line between 60-61

Comment (AdamCoops) [inline test/controllers/DeclarationDetailControllerSpec.scala:1]: new line between 79-80

Comment (AdamCoops) [inline test/controllers/DeclarationDetailControllerSpec.scala:1]: new line between 98 - 99

Comment (AdamCoops) [inline test/controllers/DeclarationDetailControllerSpec.scala:1]: new line between 158 - 159

Comment (AdamCoops) [inline test/controllers/DeclarationDetailControllerSpec.scala:1]: I think we need some spacing around the when's on 209 - 215

Comment (AdamCoops) [inline test/controllers/DeclarationDetailControllerSpec.scala:1]: same for 370 -375

File: test/views/components/H1InnerSpec.scala [hunk 6]
```diff
@@ -0,0 +1,102 @@
...
+
+        private val h1InnerComponentWithEmptyInnerMsgAsText = h1InnerComponentWithEmptyInnerMsg.text()
+
+        h1InnerComponentWithEmptyInnerMsgAsText mustBe messages(app)(msgKey, emptyString)
+        h1InnerComponentWithEmptyInnerMsg.select("h1").attr("class") mustBe classesValue
+        h1InnerComponentWithEmptyInnerMsg.select("h1").hasAttr("id") mustBe false
+      }
+
+      "it handles empty class string" in new Setup {
+        private val h1InnerComponentWithEmptyClassAsText = h1InnerComponentWithEmptyClass.text()
```
Comment (AdamCoops) [inline test/views/components/H1InnerSpec.scala:68]: the other test have space between the name and the first line, lets either add one here or remove those.

File: conf/messages.en [hunk 7]
```diff
@@ -94,6 +94,7 @@ cf.cash-account.detail.try-again-later=Try again later.
 cf.cash-account-detail.back-to-cash-account=Back to Cash account
 cf.cash-account-detail.exceeded-threshold = There are too many transactions from the last 6 months to display.
 cf.cash-account.detail.declaration.title=Declaration transaction
+cf.cash-account.detail.declaration.search-title=Search results for {0}
```
[the remark this answers is not in the document]
Comment (Neozxc) [reply conf/messages.en:97]: Yes once Sujay is back from Annual Leave. I have messaged Harsha on this to see if we have somewhere else.

File: app/models/response/CashTransactionSearchResponse.scala
Comment (tyagikumarvineet) [inline app/models/response/CashTransactionSearchResponse.scala:1]: Could you please add tests at least for for CashAccountTransactionSearchResponseDetail, reads and writes?

File: app/models/request/CashTransactionsSearchRequest.scala
Comment (tyagikumarvineet) [inline app/models/request/CashTransactionsSearchRequest.scala:1]: Could you write test cases for the classes in this file?

File: test/controllers/DeclarationDetailControllerSpec.scala [hunk 8]
```diff
@@ -104,35 +136,202 @@ class DeclarationDetailControllerSpec extends SpecBase {
...
 
     val mockCustomsFinancialsApiConnector: CustomsFinancialsApiConnector = mock[CustomsFinancialsApiConnector]
+    val mockErrorHandler: ErrorHandler = mock[ErrorHandler]
+    val mockCashAccountUtils: CashAccountUtils = mock[CashAccountUtils]
+    val mockIdentifierAction: IdentifierAction = mock[IdentifierAction]
+    val mockEmailAction: EmailAction = mock[EmailAction]
+    val mockAppConfig: AppConfig = mock[AppConfig]
+    val mockNoTransactionsView: cash_transactions_no_result = mock[cash_transactions_no_result]
+
+    val controller = new DeclarationDetailController(
```
Comment (tyagikumarvineet) [inline test/controllers/DeclarationDetailControllerSpec.scala:300]: Could you remove this and related imports as it is not being used?

File: test/controllers/DeclarationDetailControllerSpec.scala [hunk 9]
```diff
@@ -104,35 +136,202 @@ class DeclarationDetailControllerSpec extends SpecBase {
...
+      stubMessagesControllerComponents(),
+      mock[cash_account_declaration_details],
+      mock[cash_account_declaration_details_search],
+      mockCashAccountUtils,
+      mockNoTransactionsView
+    )(ExecutionContext.global, mockAppConfig)
+
+    val taxTypes: Seq[TaxType] = Seq(TaxType(reasonForSecurity = Some("Reason"), taxTypeID = "50", amount = hundred))
+
+    val declaration: Declaration = Declaration(
```
Comment (tyagikumarvineet) [inline test/controllers/DeclarationDetailControllerSpec.scala:314]: Could you remove this and related imports as it is not being used?

File: app/views/cash_account_declaration_details_search.scala.html
Comment (tyagikumarvineet) [inline app/views/cash_account_declaration_details_search.scala.html:1]: Could you add a test class for this view?

File: app/controllers/CashAccountV2Controller.scala [hunk 10]
```diff
@@ -86,8 +87,13 @@ class CashAccountV2Controller @Inject()(authenticate: IdentifierAction,
       }
   }
 
-  def onSubmit(page: Option[Int]): Action[AnyContent] = Action.async {
-    Future.successful(NotImplemented("Needs to be implemented"))
+  def onSubmit(page: Option[Int]): Action[AnyContent] = (authenticate andThen verifyEmail).async { implicit request =>
```
Comment (tyagikumarvineet) [inline app/controllers/CashAccountV2Controller.scala:90]: We need to handle the form error scenarios as well. Right now we are directing to Page not found page and not a meaningful message or a page. Could you discuss this with Harsha and get a ticket raised for this?

File: app/connectors/CustomsFinancialsApiConnector.scala [hunk 11]
```diff
@@ -141,6 +140,37 @@ class CustomsFinancialsApiConnector @Inject()(httpClient: HttpClientV2,
...
+    val request = CashAccountTransactionSearchRequestDetails(
+      can, ownerEORI, searchType, declarationDetails, cashAccountPaymentDetails
+    )
+
+    logger.info(s"SENDING REQUEST TO THE BACKEND: ${Json.toJson(request)}")
+
+    httpClient.post(url"$retrieveCashAccountStatementSearchUrl")
+      .withBody[CashAccountTransactionSearchRequestDetails](request)
+      .execute[CashAccountTransactionSearchResponseDetail].map(Right(_))
+  }.recover {
```
[the remark this answers is not in the document]
Comment (Neozxc) [reply app/connectors/CustomsFinancialsApiConnector.scala:158]: As discussed will be a part of another ticket to handle error types correctly

File: app/controllers/CashAccountV2Controller.scala (hunk 10 above)
Comment (Neozxc) [reply app/controllers/CashAccountV2Controller.scala:89]: As discussed will be a part of another ticket, as there's no error validation now and too much work for this ticket to handle all the errors

File: test/models/request/CashTransactionsSearchRequestSpec.scala [hunk 12]
```diff
@@ -0,0 +1,79 @@
...
+package models.request
+
+import play.api.libs.json.{JsSuccess, Json}
+import utils.SpecBase
+
+class CashTransactionsSearchRequestSpec extends SpecBase {
+
+  "CashAccountTransactionSearchRequestDetailsSpec" must {
+
+    "populate correctly" in new Setup {
```
Comment (tyagikumarvineet) [inline test/models/request/CashTransactionsSearchRequestSpec.scala:26]: Please remove this test as its only populating the class and of no use. Its inherent functionality

File: app/models/request/CashTransactionsSearchRequest.scala
Comment (tyagikumarvineet) [inline app/models/request/CashTransactionsSearchRequest.scala:1]: Remove rest of the class and object apart from CashAccountTransactionSearchRequestDetails and related classes/object. We need only CashAccountTransactionSearchRequestDetails for ACC44 call from frontend

File: test/models/response/CashTransactionSearchResponseSpec.scala [hunk 13]
```diff
@@ -0,0 +1,219 @@
...
+
+import play.api.libs.json.{JsSuccess, Json}
+import utils.SpecBase
+import play.api.libs.json._
+
+class CashTransactionSearchResponseSpec extends SpecBase {
+
+  "CashAccountTransactionSearchResponseDetailSpec" must {
+
+    "populate correctly with all the fields" in new Setup {
```
Comment (tyagikumarvineet) [inline test/models/response/CashTransactionSearchResponseSpec.scala:27]: Please remove this and tests starts at line 37, 47 as these are inherent functionality. We should only have tests for Json Reads and Writes

File: app/viewmodels/DeclarationDetailSearchViewModel.scala [hunk 14]
```diff
@@ -0,0 +1,156 @@
...
+import models.response.DeclarationSearch
+import play.api.i18n.Messages
+import play.twirl.api.Html
+import uk.gov.hmrc.govukfrontend.views.Aliases.{HtmlContent, SummaryList, SummaryListRow, Text}
+import uk.gov.hmrc.govukfrontend.views.viewmodels.summarylist.{Key, Value}
+import utils.Utils.{emptyH1InnerComponent, emptyH2InnerComponent, emptyString}
+
+import java.time.LocalDate
+
+case class DeclarationDetailSearchViewModel(searchInput: String,
```
[the remark this answers is not in the document]
Comment (Neozxc) [reply app/viewmodels/DeclarationDetailSearchViewModel.scala:30]: searchInput and account is used for the headers removed eori

File: test/models/response/CashTransactionSearchResponseSpec.scala (hunk 13 above)
Comment (tyagikumarvineet) [reply test/models/response/CashTransactionSearchResponseSpec.scala:27]: Could you please remove test at line 27?

File: app/models/request/CashTransactionsSearchRequest.scala
Comment (tyagikumarvineet) [reply app/models/request/CashTransactionsSearchRequest.scala:1]: Could you please remove object CashAccountTransactionSearchRequest as it is not being used?

File: app/viewmodels/DeclarationDetailSearchViewModel.scala (hunk 14 above)
[the remark this answers is not in the document]
Comment (tyagikumarvineet) [reply app/viewmodels/DeclarationDetailSearchViewModel.scala:30]: I meant remove from case class as those are not being used on the view

**Review by tyagikumarvineet — APPROVED** [review]

**Pull request merged**